pkg/cli/misc.go

```diff
@@ -29,8 +31,26 @@ import (
 
 // getServiceClients authenticates against OpenStack and returns respective ServiceClients
 // for Keystone and Limes.
-func getServiceClients() (identityV3, limesV1 *gophercloud.ServiceClient) {
-	provider, err := clientconfig.AuthenticatedClient(nil)
+func getServiceClients(OS OpenStackVariables) (identityV3, limesV1 *gophercloud.ServiceClient) {
+	authOpts := gophercloud.AuthOptions{
+		IdentityEndpoint: OS.AuthURL,
+		Username:         OS.Username,
+		Password:         os.Getenv("OS_PASSWORD"),
+		DomainName:       OS.UserDomainName,
+		AllowReauth:      true,
+		Scope: &gophercloud.AuthScope{
+			ProjectName: OS.ProjectName,
+			DomainName:  OS.ProjectDomainName,
+		},
+	}
+
+	var provider *gophercloud.ProviderClient
+	var err error
+	if OS == (OpenStackVariables{}) {
+		provider, err = clientconfig.AuthenticatedClient(nil)
+	} else {
+		provider, err = openstack.AuthenticatedClient(authOpts)
+	}
```

Comment:
The problem with this implementation is that it's all-or-nothing: You can either get everything from env, or nothing from env. The python-openstackclient, afaik, allows to use commandline options to override single env vars. We can do that as well like this:

```go
if OS.AuthURL != "" {
  os.Setenv("OS_AUTH_URL", OS.AuthURL)
}
//... and so on for all other fields in the OS struct ...
provider, err := clientconfig.AuthenticatedClient(nil)
```

This would also simplify the diff on this branch dramatically since you can do the setenv calls at the beginning of `func main()`, right after `app.Parse()`, and you won't have to carry those auth options down into all the subcommands.

Reply:
I see, I thought it was all-or-nothing that is why I went with this approach. In that case, we could also do:

```
	opts := &clientconfig.ClientOpts{
		AuthInfo: &clientconfig.AuthInfo{
			AuthURL:           OS.AuthURL,
			Username:          OS.Username,
			UserDomainName:    OS.UserDomainName,
			ProjectName:       OS.ProjectName,
			ProjectDomainName: OS.ProjectDomainName,
		},
	}

	provider, err := clientconfig.AuthenticatedClient(opts)
```

and that will also override the options, in case one is provided, right?